### DISCUSSION TITLE
Input emulation on Wayland via `libei` and `RemoteDesktop` portal

### DISCUSSION THREAD

**Discussion by MayeulC** [rustdesk/rustdesk#4515] [Feature Request]
This is an early feature request, as the necessary parts are not all supported by various upstream projects. The goal is to address the following issue in the README:

> Wayland does not seem to provide any API for sending keypresses to other windows.

[libei](https://libinput.pages.freedesktop.org/libei/api/index.html) provides an API for application to talk to Wayland compositors, and send emulated input event. It's basically mimicking the libinput<->compositor connection, but for emulated events (it goes `libei (client) -> libeis (server/compositor)`).

[This page](https://gitlab.freedesktop.org/libinput/libei/-/issues/1) tracks implementation progress across necessary parts of the software stack.

The RemoteDesktop portal should be able to provide applications that ask for it a connection to the `eis` server-part in the compositor, via [this addition](https://github.com/flatpak/xdg-desktop-portal/pull/762).

Once the pieces land, this should allow applications such as RustDesk to implement unprivileged (no need for root) remote access through various Wayland APIs, even if packaged as a sandboxed flatpak.

**Comment by nielsdg** [comment]
Note that GNOME 45 (released last September) and the latest Xwayland release (23.2) have libei support implemented already, so it should be possible to test for example on a Fedora 39 machine.

**Reply by nielsdg** [reply]
Note however that you don't need libei necessarily for the RemoteDesktop portal. If it's not available, you can still fall back on the v1 methods to explicitly create e.g. virtual key presses
